### PR TITLE
Query: Translate new overloads of Trim/TrimStart/TrimEnd in netcoreapp2.0

### PR DESCRIPTION
**Pull request by smitpatel**
Part of #8157

### CODE REVIEW

**Comment by smitpatel** [comment]
Issue here is:
netstandard2.0 has `Trim(params char)` method.
net framework implements above and adds overload `Trim()`
netcoreapp2.0 went beyond that and implemented above method & overload and also added another overload `Trim(char)`.
Therefore we were failing to resolve to correct overload while translating methodcall.

Same goes for TrimStart & TrimEnd.

File: src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringTrimEndTranslator.cs [hunk 1]
```diff
@@ -23,11 +26,14 @@ public class SqlServerStringTrimEndTranslator : IMethodCallTranslator
         /// </summary>
         public virtual Expression Translate(MethodCallExpression methodCallExpression)
         {
-            if (_methodInfo.Equals(methodCallExpression.Method)
+            if (_methodInfoWithoutArgs.Equals(methodCallExpression.Method)
+                || _methodInfoWithCharArrayArg.Equals(methodCallExpression.Method)
+                && methodCallExpression.Arguments.Count == 1
```
Comment (maumar) [inline src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringTrimEndTranslator.cs]: redundant

File: src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringTrimStartTranslator.cs [hunk 2]
```diff
@@ -23,11 +26,14 @@ public class SqlServerStringTrimStartTranslator : IMethodCallTranslator
         /// </summary>
         public virtual Expression Translate(MethodCallExpression methodCallExpression)
         {
-            if (_methodInfo.Equals(methodCallExpression.Method)
+            if (_methodInfoWithoutArgs.Equals(methodCallExpression.Method)
+                || _methodInfoWithCharArrayArg.Equals(methodCallExpression.Method)
+                && methodCallExpression.Arguments.Count == 1
```
Comment (maumar) [inline src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringTrimStartTranslator.cs]: redundant

File: src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringTrimTranslator.cs [hunk 3]
```diff
@@ -14,18 +14,26 @@ namespace Microsoft.EntityFrameworkCore.Query.ExpressionTranslators.Internal
...
         /// <summary>
         ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
         public virtual Expression Translate(MethodCallExpression methodCallExpression)
         {
-            if (_methodInfo.Equals(methodCallExpression.Method))
+            if (_methodInfoWithoutArgs.Equals(methodCallExpression.Method)
+                || _methodInfoWithCharArrayArg.Equals(methodCallExpression.Method)
+                && methodCallExpression.Arguments.Count == 1
```
Comment (maumar) [inline src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringTrimTranslator.cs]: redundant

Comment (smitpatel) [inline src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringTrimTranslator.cs]: Updated

File: src/EFCore.Sqlite.Core/Query/ExpressionTranslators/Internal/SqliteStringTrimEndTranslator.cs [hunk 4]
```diff
@@ -23,13 +30,29 @@ public class SqliteStringTrimEndTranslator : IMethodCallTranslator
...
+                    if (constantValue is char singleChar)
+                    {
+                        charactersToTrim.Add(singleChar);
+                    }
+                    else if (constantValue is char[] charArray)
+                    {
+                        charactersToTrim.AddRange(charArray);
+                    }
+
+                    sqlArguments.Add(Expression.Constant(new string(charactersToTrim.ToArray()), typeof(string)));
```
Comment (maumar) [inline src/EFCore.Sqlite.Core/Query/ExpressionTranslators/Internal/SqliteStringTrimEndTranslator.cs]: this is wrong, order of the characters shouldn't matter. e.g. "alfki".TrimStart('l', 'a') should result in "fki", just like "alfki".TrimStart('a', 'l') - also add test like this

File: test/EFCore.SqlServer.FunctionalTests/Query/QuerySqlServerTest.Functions.cs [hunk 5]
```diff
@@ -1038,171 +1038,164 @@ public override void Convert_ToString()
...
+        {
+            base.IsNullOrEmpty_in_projection();
 
-        /*
-
-
-
-
-
-                public override void Substring_with_constant()
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/Query/QuerySqlServerTest.Functions.cs]: those things should be uncommented, not removed

File: src/EFCore.Sqlite.Core/Query/ExpressionTranslators/Internal/SqliteStringTrimEndTranslator.cs (hunk 4 above)
Comment (maumar) [inline src/EFCore.Sqlite.Core/Query/ExpressionTranslators/Internal/SqliteStringTrimEndTranslator.cs]: nevermind :)

File: src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringTrimStartTranslator.cs [hunk 6]
```diff
@@ -23,11 +28,13 @@ public class SqlServerStringTrimStartTranslator : IMethodCallTranslator
         /// </summary>
         public virtual Expression Translate(MethodCallExpression methodCallExpression)
         {
-            if (_methodInfo.Equals(methodCallExpression.Method)
+            if (_methodInfoWithoutArgs?.Equals(methodCallExpression.Method) == true
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringTrimStartTranslator.cs]: @bricelam - Do you have any other approach how to deal with this?
There are "extra" overloads in netcoreapp2.0. In other tfm it resolves to different overload.
At present I am checking equality if not null.

Comment (bricelam) [inline src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerStringTrimStartTranslator.cs]: This seems best to me.

**Pull request merged**